mantidproject/mantid
# Crystal field instrument resolution

Implemented dynamic boundary constraints on peak widths in fitting to a crystal field spectrum. The boundaries depend on the peak centre via a user provided function.

**To test:**

Run this script

```
from CrystalField import CrystalField, CrystalFieldFit
from CrystalField.fitting import makeWorkspace
origin = CrystalField('Ce', 'C2v', B20=0.37, B22=3.97, B40=-0.0317, B42=-0.116, B44=-0.12,
              Temperature=44.0, FWHM=1.0)
origin.peaks.param[0]['FWHM'] = 1.01
origin.peaks.param[1]['FWHM'] = 1.1
origin.peaks.param[2]['FWHM'] = 1.8

x, y = origin.getSpectrum()
ws = makeWorkspace(x, y)

cf = CrystalField('Ce', 'C2v', B20=0.37, B22=3.97, B40=-0.0317, B42=-0.116, B44=-0.12,
                  Temperature=44.0, FWHM=1.0, ResolutionModel=([0, 70], [1, 2]), FWHMVariation=0.1)

fit = CrystalFieldFit(Model=cf, InputWorkspace=ws)
fit.fit()

res_width = lambda x: x / 70.0 + 1.0
pars = mtd['fit_Parameters']

c1 = pars.cell(38, 1)
w1 = pars.cell(39, 1)
print c1, w1, res_width(c1)

c2 = pars.cell(41, 1)
w2 = pars.cell(42, 1)
print c2, w2, res_width(c2)
```

Check that `w1` and `w2` are within `FWHMVariation` of values calculated with `res_width`.

Fixes #17517.

Release notes updated in direct_inelastic.rst

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
##### Functional Tests
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.
